IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,375 @@
+    });
+    
+    return YES;
+}
+
+- (void)dispatchCompletionBlock:(NSError *)error URL:(NSURL *)url
+{
+    // NOTE: It is possible that competition between a successful completion
+    //       and the user cancelling the authentication dialog can
+    //       occur causing this method to be called twice. The competition
+    //       cannot be blocked at its root, and so this method must
+    //       be resilient to this condition and should not generate
+    //       two callbacks.
+    [_completionLock lock];
+    
+    [MSIDChallengeHandler resetHandlers];
+    
+    if ( _completionHandler )
+    {
+        void (^completionHandler)(MSIDWebOAuth2Response *response, NSError *error) = _completionHandler;
+        _completionHandler = nil;
+        
+        MSIDWebOAuth2Response *response = nil;
+        NSError *systemError = error;
+        if (!systemError)
+        {
+            response = [MSIDWebviewAuthorization responseWithURL:url
+                                                    requestState:self.requestState
+                                                   stateVerifier:self.stateVerifier
+                                                         context:self.context
+                                                           error:&systemError];
+        }
+        
+        dispatch_async( dispatch_get_main_queue(), ^{
+            completionHandler(response, error);
+        });
+    }
+    
+    [_completionLock unlock];
+}
+
+- (void)startRequest:(NSURLRequest *)request
+{
+    [self loadRequest:request];
+    [self presentView];
+}
+
+- (void)loadRequest:(NSURLRequest *)request
+{
+    [self.webView loadRequest:request];
+}
+
+#pragma mark - WKNavigationDelegate Protocol
+
+- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler
+{
+    NSURL *requestUrl = navigationAction.request.URL;
+    NSString *requestUrlString = [requestUrl.absoluteString lowercaseString];
+    
+    MSID_LOG_VERBOSE(self.context, @"-decidePolicyForNavigationAction host: %@", [MSIDAuthority isKnownHost:requestUrl] ? requestUrl.host : @"unknown host");
+    MSID_LOG_VERBOSE_PII(self.context, @"-decidePolicyForNavigationAction host: %@", requestUrl.host);
+    
+    // Stop at the end URL.
+    if ([requestUrlString hasPrefix:[_endUrl.absoluteString lowercaseString]] ||
+        [[[requestUrl scheme] lowercaseString] isEqualToString:@"msauth"])
+    {
+        self.complete = YES;
+        
+        NSURL *url = navigationAction.request.URL;
+        [self webAuthDidCompleteWithURL:url];
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+    if ([requestUrlString isEqualToString:@"about:blank"])
+    {
+        decisionHandler(WKNavigationActionPolicyAllow);
+        return;
+    }
+    
+    if ([[[requestUrl scheme] lowercaseString] isEqualToString:@"browser"])
```

Comment:
this is very AAD specific logic. One of the requirements was to separate AAD specific logic from "base logic"

Reply:
addressed by having a subclass `MSIDAADOAuthEmbeddedWebviewController`.